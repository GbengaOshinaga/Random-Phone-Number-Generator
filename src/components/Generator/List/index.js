import React from 'react'
import { hasReachedBottom, getNumberOfItemsToRender } from '../../../utils'

class List extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      numberToRender: 50
    }
  }

  componentDidMount() {
    document.addEventListener('scroll', this.onEndReached)
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.onEndReached)
  }

  onEndReached = () => {
    const { data } = this.props
    const { numberToRender } = this.state
    const element = document.getElementById('list')
    if (hasReachedBottom(element)) {
      if (data.length > numberToRender) {
        this.setState({ numberToRender: getNumberOfItemsToRender(data.length, numberToRender) })
      }
    }
  }

  renderNumbers = () => {
    const { data } = this.props
    const { numberToRender } = this.state

    return (
      <ul className="list" id="list">
        {data.filter((number, index) => index < numberToRender).map(number => (
          <li key={number} className="item">{number}</li>
        ))}
      </ul>
    )
  }

  render() {
    const { data } = this.props

    return (
      <div>
        {data.length ? this.renderNumbers() : null}
      </div>
    )
  }
}

export default List