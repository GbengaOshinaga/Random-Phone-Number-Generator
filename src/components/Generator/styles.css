.main {
  display: flex;
  justify-content: center;
  flex-direction: column;
  font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
}

.input {
  width: 300px;
  height: 35px;
}

.top-container {
    display: flex;
    justify-content: center;
    flex-direction: column;
    background-color: darkslateblue;
    padding: 50px;
}

.input-section {
   flex-direction: row;
   justify-content: center;
   align-self: center;
}

.sort {
  align-self: center;
  margin-top: 20px;
}

.select {
  width: 150px;
  height: 30px;
}

.max-min {
  display: flex;
  align-self: center;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-top: 20px;
}

.max-num {
  margin-right: 30px;
}

.min-num {
  margin-left: 30px;
}

.max {
  color: firebrick;
  margin-right: 5px;
}

.min {
  color: darkred;
  margin-right: 5px;
}

.num {
  font-weight: bolder
}

li {
    font-weight: 400;
    margin-top: 20px;
    margin-bottom: 20px;
}

.download-button {
  width: 120px;
  align-self: center;
  margin-top: 20px;
}
