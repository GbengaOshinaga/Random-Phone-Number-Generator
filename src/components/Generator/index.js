import React from 'react'
import FileSaver from 'file-saver'
import Button from '../Button'
import List from './List'
import ScrollToTopButton from '../ScrollToTopButton'
import { 
  generateNumbers,
  getMaxNumber,
  getMinNumber,
  sortAscending,
  sortDescending
} from '../../utils'
import './styles.css'

class Generator extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      numberOfGenerations: 0,
      numbers: [],
      sortOrder: 'ascending',
      showScrollToTop: false
    }
  }

  componentDidMount() {
    document.addEventListener('scroll', this.onScroll)
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.onScroll)
  }

  onScroll = () => {
    const { body, documentElement } = document

    if (body.scrollTop > 40 || documentElement.scrollTop > 40) {
      this.setState({ showScrollToTop: true })
    } else {
      this.setState({ showScrollToTop: false })
    }
  }

  onTextChange = event => {
    const { value } = event.target
    !isNaN(value) && this.setState({ numberOfGenerations: value })
  }

  onSortChange = event => {
    const { numbers } = this.state
    const { value } = event.target

    const sortedNumbers = value === "ascending" ? sortAscending(numbers) : sortDescending(numbers)

    this.setState({ sortOrder: value, numbers: sortedNumbers })
  }

  onGenerateButtonClick = () => {
    const { numberOfGenerations, sortOrder } = this.state
    const numbers = generateNumbers(numberOfGenerations)
    const sortedNumbers = sortOrder === 'ascending' ? sortAscending(numbers) : sortDescending(numbers)
    this.setState({ numbers: sortedNumbers })
  }

  onDownloadNumbersClick = () => {
    const { numbers } = this.state
    const numbersString = numbers.join(', ')

    const blob = new Blob([numbersString], { type: 'text/plain;charset=utf-8' })
    FileSaver.saveAs(blob, 'numbers.txt')
  }

  renderInputSection = () => {
    const { numbers } = this.state

    return (
      <div className="top-container">
        <div className="input-section">
          <input
            className="input"
            autoFocus
            type="number"
            placeholder="Enter number of phone numbers to generate"
            onChange={this.onTextChange}
          />
          <Button id="generate-button" onClick={this.onGenerateButtonClick}>Generate</Button>
        </div>
        {numbers.length 
          ? 
          (
            <Button
             id="download-button"
             className="download-button"
             onClick={this.onDownloadNumbersClick}>
              Download Numbers
             </Button>
          )
          : null}
      </div>
    )
  }

  renderMaxAndMin = () => {
    const { numbers } = this.state
    return numbers.length > 2 ? (
      <div className="max-min">
        <div className="max-num">
          <span className="max">Max Num:</span>
          <span className="num">{`0${getMaxNumber(numbers)}`}</span>
        </div>
        <div className="min-num">
          <span className="min">Min Num:</span> 
          <span className="num">{`0${getMinNumber(numbers)}`}</span>
        </div>
      </div>
    ) : null
  }

  renderSortOptions = () => {
    return (
      <div className="sort">
        <select className="select" onChange={this.onSortChange}>
          <option value="ascending">Ascending</option>
          <option value="descending">Descending</option>
        </select>
      </div>
    )
  }

  renderList = () => {
    const { numbers } = this.state
    return <List data={numbers} />
  }

  renderScrollToTopButton = () => {
    const { showScrollToTop } = this.state
    return showScrollToTop ? 
    (
      <ScrollToTopButton 
        id="scroll-button"
        scrollStepInPx="80"
        delayInMs="3"
      />
      ) : null
  }

  render () {
    return (
      <div className="main">
        {this.renderInputSection()}
        {this.renderSortOptions()}
        {this.renderMaxAndMin()}
        {this.renderList()}
        {this.renderScrollToTopButton()}
      </div>
    )
  }
}

export default Generator
