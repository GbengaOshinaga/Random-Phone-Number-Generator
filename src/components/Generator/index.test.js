import React from 'react'
import { mount } from 'enzyme'
import FileSaver from 'file-saver'
import renderer from 'react-test-renderer'
import Generator from './index'

it('renders correctly', () => {
  let tree = renderer.create(<Generator />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe('componentDidMount', () => {
  it('should register listener', () => {
    const map = {}
    document.addEventListener = jest.fn((event, callback) => {
      map[event] = callback
    })
    const generator = mount(<Generator />)
    generator.instance().onScroll = jest.fn()
    generator.instance().componentDidMount()
    map.scroll({ key: 'scroll' })
    expect(generator.instance().onScroll).toHaveBeenCalled()
  })
})

describe('onTextChange', () => {
  it('should set state if text is a number', () => {
    const generator = mount(<Generator />)
    const inputProps = generator.find('input').props()
    const event = { target: { value: '1' } };
    inputProps.onChange(event)
    expect(generator.instance().state.numberOfGenerations).toEqual('1')
  })
  it('should not set state if text is not a number', () => {
    const generator = mount(<Generator />)
    const inputProps = generator.find('input').props()
    const event = { target: { value: 'a' } };
    inputProps.onChange(event)
    expect(generator.instance().state.numberOfGenerations).toEqual(0)
  })
})

describe('onSortChange', () => {
  it('should sort numbers in ascending order', () => {
    const generator = mount(<Generator />)
    const optionProps = generator.find('select').props()
    const event = { target: { value: 'ascending' } };
    generator.instance().state.numbers = [3,5,6,2,1]
    optionProps.onChange(event)
    expect(generator.instance().state.numbers).toEqual([1,2,3,5,6])
  })
  it('should sort numbers in descending order', () => {
    const generator = mount(<Generator />)
    const optionProps = generator.find('select').props()
    const event = { target: { value: 'descending' } };
    generator.instance().state.numbers = [3,5,6,2,1]
    optionProps.onChange(event)
    expect(generator.instance().state.numbers).toEqual([6,5,3,2,1])
  })
})

describe('onGenerateButtonClick', () => {
  it('should generate numbers', () => {
    const generator = mount(<Generator />)
    generator.instance().state.numberOfGenerations = 2
    const buttonProps = generator.find('button#generate-button').props()
    buttonProps.onClick()
    expect(generator.instance().state.numbers.length).toEqual(2)
  })
})

describe('onDownloadNumbersClicked', () => {
  it('should generate numbers', () => {
    const generator = mount(<Generator />)
    generator.instance().state.numbers = [3,5,6,2,1]
    FileSaver.saveAs = jest.fn()
    generator.instance().onDownloadNumbersClick()
    expect(FileSaver.saveAs).toHaveBeenCalled()
  })
})

describe('onScroll', () => {
  it('should update state accordingly', () => {
    const generator = mount(<Generator />)
    document.body.scrollTop = 100
    document.documentElement.scrollTop = 100
    generator.instance().onScroll()
    expect(generator.instance().state.showScrollToTop).toEqual(true)

    document.body.scrollTop = 10
    document.documentElement.scrollTop = 10
    generator.instance().onScroll()
    expect(generator.instance().state.showScrollToTop).toEqual(false)

  })
})
