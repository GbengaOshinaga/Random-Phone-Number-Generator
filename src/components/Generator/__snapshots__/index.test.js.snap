// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="main"
>
  <div
    className="top-container"
  >
    <div
      className="input-section"
    >
      <input
        autoFocus={true}
        className="input"
        onChange={[Function]}
        placeholder="Enter number of phone numbers to generate"
        type="number"
      />
      <button
        className="button undefined"
        id="generate-button"
        onClick={[Function]}
        type="button"
      >
        Generate
      </button>
    </div>
  </div>
  <div
    className="sort"
  >
    <select
      className="select"
      onChange={[Function]}
    >
      <option
        value="ascending"
      >
        Ascending
      </option>
      <option
        value="descending"
      >
        Descending
      </option>
    </select>
  </div>
  <div />
</div>
`;
